cpp/src/arrow/json/parser.cc

```diff
@@ -610,12 +627,22 @@ class HandlerBase : public BlockParser,
   }
 
   bool RawNumber(const char* data, rj::SizeType size, ...) {
-    status_ = AppendScalar<Kind::kNumber>(builder_, std::string_view(data, size));
+    if (builder_.kind == Kind::kNumberOrString) {
+      status_ =
+          AppendScalar<Kind::kNumberOrString>(builder_, std::string_view(data, size));
+    } else {
```

Comment:
Instead of doing this check here, how about doing it in `AppendScalar`, e.g.:
```c++
  template <Kind::type kind>
  Status AppendScalar(BuilderPtr builder, std::string_view scalar) {
    bool kind_changed;
    if (kind == Kind::kNumber || kind == Kind::kString) {
      kind_changed = builder.kind != kind && builder.kind != Kind::kNumberOrString;
    } else {
      kind_changed = builder.kind != kind;
    }
    if (ARROW_PREDICT_FALSE(kind_changed)) {
      return IllegallyChangedTo(kind);
    }
```

Reply:
I'm not sure if we want this. This way we need add two if checks. One is for `kind_changed` as you write here. The other is when doing `Cast<kind>(builder)` later. We need to do
```cpp
if (builder.kind == Kind::kNumberOrString) {
  RETURN_NOT_OK(Cast<kind>(builder)->Append(index, value_length));
} else {
  RETURN_NOT_OK(Cast<Kind::kNumberOrString>(builder)->Append(index, value_length));
}
```
I think this way the code is more complex. What do you think?